### ISSUE TITLE
如果训练时不走flash路径，是否会出现出现问题？

### ISSUE THREAD

**Issue by MonicaJitao** [jingyaogong/minimind#650]
在/minimind/model/model_minimind.py中，Attention类如下
```python
class Attention(nn.Module):
    def __init__(self, args: MiniMindConfig):
        """
        搭建维度关系和线性层
        """
        super().__init__() # 调用父类初始化方法
        self.num_key_value_heads = args.num_attention_heads if args.num_key_value_heads is None else args.num_key_value_heads # 如果 config 没设 KV 头数，就默认等于 Q 头数（退化为普通 MHA）
        assert args.num_attention_heads % self.num_key_value_heads == 0
        self.n_local_heads = args.num_attention_heads
        self.n_local_kv_heads = self.num_key_value_heads
        self.n_rep = self.n_local_heads // self.n_local_kv_heads
        self.head_dim = args.hidden_size // args.num_attention_heads
        self.q_proj = nn.Linear(args.hidden_size, args.num_attention_heads * self.head_dim, bias=False) # 线性层，把 hidden_size 映射到 n_h * head_dim
        self.k_proj = nn.Linear(args.hidden_size, self.num_key_value_heads * self.head_dim, bias=False) # 线性层，把 hidden_size 映射到 n_kv * head_dim
        self.v_proj = nn.Linear(args.hidden_size, self.num_key_value_heads * self.head_dim, bias=False) # 线性层，把 hidden_size 映射到 n_kv * head_dim
        self.o_proj = nn.Linear(args.num_attention_heads * self.head_dim, args.hidden_size, bias=False) # 线性层，把 n_h * head_dim 映射到 hidden_size
        self.attn_dropout = nn.Dropout(args.dropout) # 注意力dropout
        self.resid_dropout = nn.Dropout(args.dropout) # 残差dropout
        self.dropout = args.dropout
        self.flash = hasattr(torch.nn.functional, 'scaled_dot_product_attention') and args.flash_attn # 是否使用 Flash Attention
        # print("WARNING: using slow attention. Flash Attention requires PyTorch >= 2.0")

    def forward(self,
                x: torch.Tensor,
                position_embeddings: Tuple[torch.Tensor, torch.Tensor],  # 修改为接收 cos 和 sin
                past_key_value: Optional[Tuple[torch.Tensor, torch.Tensor]] = None,
                use_cache=False,
                attention_mask: Optional[torch.Tensor] = None):
        bsz, seq_len, _ = x.shape
        xq, xk, xv = self.q_proj(x), self.k_proj(x), self.v_proj(x)
        xq = xq.view(bsz, seq_len, self.n_local_heads, self.head_dim) # 重排数据维度：[B, T, n_h*d] → reshape → [B, T, n_h, d]
        xk = xk.view(bsz, seq_len, self.n_local_kv_heads, self.head_dim)
        xv = xv.view(bsz, seq_len, self.n_local_kv_heads, self.head_dim)

        cos, sin = position_embeddings
        xq, xk = apply_rotary_pos_emb(xq, xk, cos[:seq_len], sin[:seq_len])

        # kv_cache实现
        if past_key_value is not None:
            xk = torch.cat([past_key_value[0], xk], dim=1)
            xv = torch.cat([past_key_value[1], xv], dim=1)
        past_kv = (xk, xv) if use_cache else None

        xq, xk, xv = (
            xq.transpose(1, 2), # [B, T, n_h, d] → [B, n_h, T, d]（注意力计算常用布局，按头分组，每组都有头的整个句子，更符合注意力“每个头读完整篇文章做计算”的直觉）
            repeat_kv(xk, self.n_rep).transpose(1, 2),
            repeat_kv(xv, self.n_rep).transpose(1, 2)
        )

        if self.flash and seq_len > 1 and (attention_mask is None or torch.all(attention_mask == 1)):
            output = F.scaled_dot_product_attention(xq, xk, xv, dropout_p=self.dropout if self.training else 0.0, is_causal=True)
        else:
            scores = (xq @ xk.transpose(-2, -1)) / math.sqrt(self.head_dim)
            #  构造 [T, T] 的 mask，上三角为 -inf，unsqueeze(0).unsqueeze(0) 变成 [1,1,T,T]
            scores = scores + torch.triu(
                torch.full((seq_len, seq_len), float("-inf"), device=scores.device),
                diagonal=1
            ).unsqueeze(0).unsqueeze(0)  

            if attention_mask is not None:
                extended_attention_mask = attention_mask.unsqueeze(1).unsqueeze(2) # [B,T] → [B,1,1,T]
                extended_attention_mask = (1.0 - extended_attention_mask) * -1e9 # mask 值为 0 或 1，因此这是取反后，原本 0 的位置变成 -inf，加到 attention logits 后，softmax ≈ 0
                scores = scores + extended_attention_mask

            scores = F.softmax(scores.float(), dim=-1).type_as(xq) # 在 fp32 下算 softmax 更稳定（精度更高，避免数值上溢/下溢），然后转换回 xq 的 dtype
            scores = self.attn_dropout(scores)
            output = scores @ xv

        output = output.transpose(1, 2).reshape(bsz, seq_len, -1)
        output = self.resid_dropout(self.o_proj(output)) # [B,n_h,T,T_total] @ [B,n_h,T_total,d] → [B,n_h,T,d]
        return output, past_kv
```
如果不走```if self.flash and seq_len > 1 and (attention_mask is None or torch.all(attention_mask == 1)):```条件，则casual mask 用的是 [seq_len, seq_len]，而当 past_key_value 拼接后，xk 的长度其实是 T_total = T_past + T。是否需要构造 [T, T_total] 的casual mask（确保当前 chunk 的每个位置只看历史+当前已到的位置）？

**Comment by jingyaogong** [comment]
> “训练时不走flash路径，是否会出现出现问题？”

此时 total_len == seq_len，切片覆盖整个矩阵，应用标准的 Causal Mask


> 如果不走if self.flash and seq_len > 1 and (attention_mask is None or torch.all(attention_mask == 1)):条件，则casual mask 用的是 [seq_len, seq_len]，而当 past_key_value 拼接后，xk 的长度其实是 T_total = T_past + T。是否需要构造 [T, T_total] 的casual mask（确保当前 chunk 的每个位置只看历史+当前已到的位置）？

不知你的title和这里的问题有无关联？“past_key_value存在” 应该是inference阶段：

* 历史部分（Total_Len 的前 T_past 列）：这部分对应 past_key_value，对于当前的 Query 来说都是“过去”的信息，因此不需要 再Mask（天然就全可见）
* 当前部分（Total_Len 的后 seq_len 列）：这部分对应当前的输入，需要保证因果性（不能看未来），所以加上了 [seq_len, seq_len] 的上三角 Mask

没有问题

**Comment by MonicaJitao** [comment]
感谢你的及时回复和耐心教导，该项目真的帮助我很多，谢谢你
